### PR TITLE
Option to turn arbitrary input into sygus

### PR DESCRIPTION
**Pull request by ajreynol**
Preprocessing pass that turns an arbitrary (e.g. smt2) input into a sygus conjecture, which is helpful for Horn clause solving.

This includes improvements to the robustness of the sygus solver.

### CODE REVIEW

File: src/theory/quantifiers/sygus/ce_guided_single_inv.cpp [hunk 1]
```diff
@@ -14,6 +14,8 @@
  **/
 #include "theory/quantifiers/sygus/ce_guided_single_inv.h"
 
+#include "theory/quantifiers/sygus/sygus_grammar_cons.h"
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/ce_guided_single_inv.cpp]: this is already included in the block below, remove it

File: src/theory/quantifiers/sygus_inference.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of sygus_inference
+ **/
+
+#include "theory/quantifiers/sygus_inference.h"
+#include "theory/quantifiers/quantifiers_attributes.h"
+
+using namespace std;
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_inference.cpp]: kill

File: src/theory/quantifiers/sygus_inference.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+          if (std::find(free_functions.begin(), free_functions.end(), op)
+              == free_functions.end())
+          {
+            free_functions.push_back(op);
+          }
+        }
+        else if (cur.getKind() == FORALL)
+        {
+          Trace("sygus-infer")
+              << "...fail: non-top-level quantifier." << std::endl;
```
Comment (HanielB) [inline src/theory/quantifiers/sygus_inference.cpp]: Add an issue to be robust to such cases. This function could prenex the formula as well as assume that skolemization has been performed (another issue to allow this inference to happen after preprocessing).

Comment (ajreynol) [reply src/theory/quantifiers/sygus_inference.cpp]: We are now robust to preskolemization and prenexing after my latest commit.

**Pull request merged**